pkg/workloads/containerd/watcher.go

```diff
@@ -49,36 +50,63 @@ const (
 	maxRetries = 3
 )
 
-// EnableDockerEventListener watches for docker events. Performs the plumbing for the
+func shortContainerID(id string) string {
+	return id[:10]
+}
+
+// EnableEventListener watches for docker events. Performs the plumbing for the
 // containers started or dead.
-func (d *Daemon) EnableDockerEventListener(since time.Time) error {
+func EnableEventListener() error {
+	since := time.Now()
+	syncWithRuntime()
+
 	eo := dTypes.EventsOptions{Since: strconv.FormatInt(since.Unix(), 10)}
-	r, err := d.dockerClient.Events(ctx.Background(), eo)
+	r, err := dockerClient.Events(ctx.Background(), eo)
 	if err != nil {
 		return err
 	}
-	log.Debugf("Listening for docker events")
-	go d.listenForDockerEvents(r)
+
+	go listenForDockerEvents(r)
+
+	// start a go routine which periodically synchronizes containers
+	// managed by the local container runtime and checks if any of them
+	// need to be managed by Cilium. This is a fall back mechanism in case
+	// an event notification has been lost.
+	go func() {
+		for {
+			time.Sleep(syncRateDocker)
+			syncWithRuntime()
+		}
+	}()
+
+	log.Debugf("Started to listen for containerd events")
+
 	return nil
 }
 
-// SyncDocker is used by the daemon to synchronize changes between Docker and
+// syncWithRuntime is used by the daemon to synchronize changes between Docker and
 // Cilium. This includes identities, labels, etc.
-func (d *Daemon) SyncDocker() {
+func syncWithRuntime() {
 	var wg sync.WaitGroup
 
-	cList, err := d.dockerClient.ContainerList(ctx.Background(), dTypes.ContainerListOptions{All: false})
+	// FIXME: Must be synchronize with event handler
```

Comment:
Can you clarify this a little more?

Reply:
docker events and the manual sync can overlap and we could readd a container after we have removed it based on the event if the list occurred before the delete event was sent out. I'll file a github issue instead.